src/ProvidedTypes.fs

```diff
@@ -1,3022 +1,8112 @@
+                            |> (fun arr -> arr.[n - 1])
+
+                        let adaptMethod = getFastFuncType args resultType
+                        let adapted = Expr.Call(adaptMethod, [loop applicable])
+                        let invoke = adapted.Type.GetMethod("Invoke", [| for arg in args -> arg.Type |])
+                        Expr.Call(adapted, invoke, args)
+                    else
+                        (applicable, args) ||> List.fold (fun e a -> Expr.Application(e, a))
+                | _ ->
                     orig
+            and countPeelableApplications expr v n =
+                match expr with
+                // v - applicable entity obtained on the prev step
+                // \arg -> let v1 = (f arg) in rest ==> f
+                | Lambda(arg, Let(v1, Application(Var f, Var arg1), rest)) when v = f && arg = arg1 -> countPeelableApplications rest v1 (n + 1)
+                // \arg -> (f arg) ==> f
+                | Lambda(arg, Application(Var f, Var arg1)) when v = f && arg = arg1 -> n
+                | _ -> n
+            loop expr
+    #endif
+
+        member __.TranslateExpression q = transExpr q
+
+        member __.TranslateQuotationToCode qexprf (paramNames: string[]) (argExprs: Expr[]) =
+            // Use the real variable names instead of indices, to improve output of Debug.fs
+            // Add let bindings for arguments to ensure that arguments will be evaluated
+            let vars = argExprs |> Array.mapi (fun i e -> Quotations.Var(paramNames.[i], e.Type))
+            let expr = qexprf ([for v in vars -> Expr.Var v])
+
+            let pairs = Array.zip argExprs vars
+            let expr = Array.foldBack (fun (arg, var) e -> Expr.LetUnchecked(var, arg, e)) pairs expr
+    #if !NO_GENERATIVE
+            let expr =
+                if isGenerated then
+                    let e1 = inlineRightPipe expr
+                    let e2 = optimizeCurriedApplications e1
+                    let e3 = inlineValueBindings e2
+                    e3
                 else
-                    e1
-            | Let(id, applicable, (Lambda(_) as body)) when n > 0 ->
-                let numberOfApplication = countPeelableApplications body id 0
-                if numberOfApplication = 0 then orig
-                elif n = 1 then Expr.Application(applicable, List.head args)
-                elif n <= 5 then
-                    let resultType =
-                        applicable.Type
-                        |> Seq.unfold (fun t ->
-                            if not t.IsGenericType then None else
-                            let args = t.GetGenericArguments()
-                            if args.Length <> 2 then None else
-                            Some (args.[1], args.[1])
-                        )
-                        |> Seq.toArray
-                        |> (fun arr -> arr.[n - 1])
-
-                    let adaptMethod = getFastFuncType args resultType
-    
```

Comment:
Is there a particular reason for `GetterCode` instead of `getterCode` here? Actually, I'm a bit confused why a lot of parameters are CamelCase. Did I miss a guideline somewhere?

Reply:
(I'm asking because from the changes it seems to be "less" source code breaking with lower case but that could be completely off)

Reply:
I am just reverting that